cdap-security/src/main/java/io/cdap/cdap/security/auth/context/AuthenticationContextModules.java

```diff
@@ -40,17 +43,40 @@
  * it is being invoked.
  */
 public class AuthenticationContextModules {
+
   /**
    * An {@link AuthenticationContext} for HTTP requests in Master. The authentication details in this context are
-   * derived from {@link SecurityRequestContext}.
+   * derived from a combination of {@link SecurityRequestContext} and the {@link SystemAuthenticationContext}.
    *
    * @see SecurityRequestContext
+   * @see SystemAuthenticationContext
    */
   public Module getMasterModule() {
-    return new AbstractModule() {
+    return new PrivateModule() {
       @Override
       protected void configure() {
-        bind(AuthenticationContext.class).to(MasterAuthenticationContext.class);
+        bind(new TypeLiteral<Class<? extends AuthenticationContext>>() { })
+          .toInstance(SystemAuthenticationContext.class);
+        bind(AuthenticationContext.class).toProvider(MasterAuthenticationContextProvider.class);
+        expose(AuthenticationContext.class);
+      }
+    };
+  }
+
+  /**
+   * Returns a Guice module that provides {@link AuthenticationContext} for workers such as preview and task workers.
+   * The authentication details in this context are derived from the {@link WorkerAuthenticationContext}.
+   *
+   * @see WorkerAuthenticationContext
+   */
+  public Module getMasterWorkerModule() {
+    return new PrivateModule() {
+      @Override
+      protected void configure() {
+        bind(new TypeLiteral<Class<? extends AuthenticationContext>>() { })
```

Comment:
This seems rather ugly, maybe named injection might be better here than type literal injection?

Reply:
It is just injecting the `Class<? extends AuthenticationContext>`. The TypeLiteral is the way to bind parameterized type. How does a named injection can change this?

Reply:
Oh sorry, to clarify, you have a binding from `AuthenticationContext` to a provider, then you bind `Class<? extends AuthenticationContext>` which is used in the provider to the actual context which is used; I was just concerned that this might be a bit confusing, but feel free to disregard as it's just a nitpick.

Reply:
The other way would be to create two provider. However, since this is all within the same class as private binding, I think it is ok with the current form.